### PR TITLE
PulseAudio sink 

### PR DESCRIPTION
**Pull request by marcust**

### CODE REVIEW

File: src/audio_sink.rs [hunk 1]
```diff
@@ -53,5 +55,78 @@ mod portaudio_sink {
...
+
+            let ss = pa_sample_spec {
+                format: PA_SAMPLE_S16LE,
+                channels: 2, // stereo
+                rate: 44100
+            };
+            
+            let s = unsafe {
+                pa_simple_new(null(),             // Use the default server.
+                              "librespot".as_ptr() as *const PtrType,  // Our application's name.
```
Comment (plietar) [inline src/audio_sink.rs]: This isn't safe, see https://doc.rust-lang.org/std/ffi/struct.CString.html for how to pass strings to C functions

File: src/audio_sink.rs [hunk 2]
```diff
@@ -53,5 +55,78 @@ mod portaudio_sink {
...
+    use audio_sink::Sink;
+    use std::io;
+    use libpulse_sys::*;
+    use std::ptr::{null, null_mut};
+    use std::mem::{transmute};
+
+    pub struct PulseAudioSink(*mut pa_simple);
+
+    #[cfg(target_arch = "arm")]
+    type PtrType = u8;
```
Comment (plietar) [inline src/audio_sink.rs]: This is `libc::c_char`

File: src/audio_sink.rs [hunk 3]
```diff
@@ -53,5 +55,78 @@ mod portaudio_sink {
...
+            Ok(())
+        }
+
+        fn stop(&self) -> io::Result<()> {
+            Ok(())
+        }
+
+        fn write(&self, data: &[i16]) -> io::Result<()> {
+            unsafe {
+                let ptr = transmute(data.as_ptr());
```
Comment (plietar) [inline src/audio_sink.rs:114]: Just use `data.as_ptr() as *const libc::c_void`

**Comment by plietar** [comment]
Thanks,
Why not use the `rust-pulse-simple` crate which exposes a safe binding on top of `pulse-sys` ?
Also this will have to wait a little before being merge, see https://github.com/plietar/librespot/issues/63#issuecomment-198417728, sorry.

**Comment by marcust** [comment]
Thanks for the feedback, I'll try to incorporate the changes (and get a better grip on the typing in general)

As you might have seen that are my first steps in Rust (and anything other than Java for a while), so I kind of got lost in the types.

The reason of not using rust-pulse-simple was that I could not figure out how to use the Playback.write method with the full buffer of data. I'm still not really sure what the type of X has to be in Playback<X> to make that work.

**Comment by plietar** [comment]
From what I understand, in rust-pulse-simple, `C` should be a `[i16; 2]`, ie an array of two i16, one for each channel.
This means data should be a `&[[i16; 2]]`. You can get one by doing `std::slice::from_raw_parts(data.as_ptr() as *const [i16; 2], data.len() / 2)`.

**Comment by marcust** [comment]
ok, I didn't know about the slice thing obviously. But I think using the calls directly makes it more obvious, at least to such noobs like me.

One of the things I don't understand is how types are actually might differ to the ones advertised:
according to your suggestion `data.as_ptr() as *const libc::c_void` should be fine, even the typing in libpulse-sys/libpulse.rs is like that, but when I try that I get a 

```
 expected `*const libc::types::common::c95::c_void`,
    found `*const libc::c_void`
(expected enum `libc::types::common::c95::c_void`,
    found enum `libc::c_void`) [E0308]
```

which puzzles me much. Where does he take this c95 from out of a sudden?

**Comment by plietar** [comment]
How did you include libc ? You should use the libc crate, not the compiler's libc feature

**Comment by plietar** [comment]
And I agree, the pulse-simple crate barely adds anything in the end, I'm fine with keeping it like that.

**Comment by marcust** [comment]
As I don' know what "the compiler's libc feature" actually means, I tried this:
https://github.com/marcust/librespot/commit/c41b178f19bb4d29ee27cf9d9c95440cd47de6bb

**Pull request closed without merging**